File: TodoList/app/src/main/java/com/callor/todo/FirstFragment.java
package com.callor.todo;

import android.os.Bundle;

import androidx.fragment.app.Fragment;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.callor.todo.databinding.FragmentFirstBinding;

/* findViewById 에 비해 viewBinding 방식 사용 장점
    1. Null PointException 에서 비교적 안전함.
    2. xml 을 직접 핸들링하기 않기 때문에 시스템에서 여유로움.
 */
public class FirstFragment extends Fragment {

    private FragmentFirstBinding binding;

    public FirstFragment() {
        // Required empty public constructor
    }
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentFirstBinding.inflate(inflater,container,false);
        View view = binding.getRoot();

        /* OnClickListener
           1. imageView 에 src 를 설정하는데 src 에 설정된 drawable 파일에는 2개의 icon 을 설정한다.
           2. 2개 중, icon state_pressed 를 true 로 설정하기
           3. imageView 에 click event 를 설정하면 버튼 터치했을 때 바뀌는 것을 볼 수 있다.
         */
        binding.heartImage.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

            }
        });

        return view;
    }

    /* 주의사항!(onDestroyView 사용이유)
       viewBinding 방식으로 Fragment 를 사용할 때
       반드시 onDestroyView() 를 정의하고
       이 메서드에서 binding 객체 변수를 null 로 초기화 해서
       메모리 누수(lick)를 방지해야 한다.
     */
    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }


}//end class

File: HelloApp/app/src/main/java/com/callor/helloapp/FirstFragment.kt
package com.callor.helloapp

import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.navigation.fragment.findNavController
import com.callor.helloapp.databinding.FragmentFirstBinding

/**
 * A simple [Fragment] subclass as the default destination in the navigation.
 */
class FirstFragment : Fragment() {

    // FragmentFirstBinding 는 fragment를 연결하는 도구이다.
    /*
    코틀린 객체 변수들은 원칙적으로 null 값 저장이 불가능하다.
    변수의 타입뒤에 ? 를 붙여주면 제한적으로 null값을 허용해 저장 가능하다
     */
    private var _binding: FragmentFirstBinding? = null


    // This property is only valid between onCreateView and onDestroyView.
    //전역변수 만드는 방법이다.
    private val binding get() = _binding!!

    /* onCreateView
     * fragment를 확장해 객체로 만드는 함수이다.
     * return type이 view? 이고
     * 3개의 매개변수를 가지는 구조이다.
     */
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {

        _binding = FragmentFirstBinding.inflate(inflater, container, false)
        return binding.root

    }
    
    /* onViewCreated
    fragment 함수를 화면에 그리는 함수
     */
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        /* firstfragment에 있는 buttonFirst를 클릭하면
        어딘가에 설정된 nav controller를 찾아서 navigate() 함수를 실행하여
        화면을 전환하는 역할을 한다.
         */
        binding.buttonFirst.setOnClickListener {
            findNavController().navigate(R.id.action_FirstFragment_to_SecondFragment)
        }
    }

    /* binding의 값이 null로 바뀌었으므로,
    이것은 화면을 지우는 함수이다.
     */
    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}